solvers

lucaferranti opened the pull request:
* Make sure solvers behave the same way for both static matrices and normal matrices (fixes #9 )

* added tests for linear solvers


I need to investigate the failure on nightly

dpsanders commented on test/runtests.jl:
```diff
@@ -1,6 +1,29 @@
+using IntervalLinearAlgebra, StaticArrays, IntervalArithmetic
 using Test
 
+int_approx(a, b, tol) =  abs(a.lo - b.lo) < tol  && abs(a.hi - b.hi) < tol
```
Int here sounds like integer. Maybe interval_approx?

lucaferranti replied:
good point, although I guess the best solution would be to open a PR to IntervalArithmetic.jl and overload `isapprox` there?

lucaferranti replied:
for the time being, I've overloaded `isapprox` here. This should be moved to `IntervalArithmetic.jl` at some point I guess

mforets commented on src/utils.jl:
```diff
@@ -13,7 +13,7 @@ that is the midpoint of A.
 function precondition(A, b)
     Ac = mid.(A)
-    return Ac\A, Ac\b
+    return inv(Ac)*A, inv(Ac)*b
```
This computes the inverse two times.

lucaferranti replied:
fixed!

lucaferranti commented:
@dpsanders @mforets this should now be ready to be merged. In a nutshell, this PR has reorganized the solvers that were already implemented, particularly

1) added pretty printing of solvers and more details in docstrings.
2) avoid using backslash internally
3) Fixed bug in formula for `enclose`
4) Overloaded `Base.isapprox` to use tolerances in iterative algorithms.

dpsanders commented on src/utils.jl:
```diff
@@ -12,36 +14,30 @@ Preconditions the interval system Ax = b by multipling by the (approximate) inve
+    C = inv(mid.(A))
+    A1 = Diagonal(ones(N)) - C*A
+    e = interval_norm(C*b)/(1 - interval_norm(A1))
     x0 = MVector{N, T}(fill(-e..e, N))
```
I think using the `ntuple` function

lucaferranti replied:
it indeed reduced the allocations a little

lucaferranti commented:
@dpsanders @mforets can this be merged?
